Document `k6/experimental/fs` module (#1419)
grafana/k6-docs: docs/sources/next/javascript-api/k6-experimental/fs/_index.md

---
title: 'fs'

> excerpt: 'k6 fs experimental API'

Reviewer: ```suggestion
description: 'k6 fs experimental API'
```

weight: 10
---

# fs

{{< docs/shared source="k6" lookup="experimental-module.md" version="<K6_VERSION>" >}}

The k6 filesystem experimental module provides a memory-efficient way to handle file interactions within your test scripts. It currently offers support for opening files, reading their content, and seeking through their content, and retrieving metadata about them.

### Memory efficiency

> One of the key advantages of the filesystem module is its memory efficiency. Unlike the traditional [open]() function, which loads a file multiple times into memory, the filesystem module reduces memory usage by loading the file as little possible, and sharing the same memory space between all VUs. This approach signally reduces the memory footprint of your test script, and lets you load and process large files without running out of memory.

Reviewer 2: >and lets you load and process large files without running out of memory.

I'm wondering if we should adjust this place to something like: 

```suggestion
One of the key advantages of the filesystem module is its memory efficiency. Unlike the traditional [open]() function, which loads a file multiple times into memory, the filesystem module reduces memory usage by loading the file as little possible, and sharing the same memory space between all VUs. This approach reduces the number of running out-of-memory errors while processing large files.
```

Since I am not sure if the new module without file-streams support allows customers to leverage the issues with the large files fully

WDYT?

Author: 👍🏻 

I've rephrased to "This approach reduces the risk of encountering out-of-memory errors, especially in load tests involving large files."

> One of the key advantages of the filesystem module is its memory efficiency. Unlike the traditional [open](https://grafana.com/docs/k6/latest/javascript-api/init-context/open/) function, which loads a file multiple times into memory, the filesystem module reduces memory usage by loading the file as little possible, and sharing the same memory space between all VUs. This approach reduces the risk of encountering out-of-memory errors, especially in load tests involving large files.

Reviewer: ```suggestion
One of the key advantages of the filesystem module is its memory efficiency. Unlike the traditional [open](https://grafana.com/docs/k6/latest/javascript-api/init-context/open/) function, which loads a file multiple times into memory, the filesystem module reduces memory usage by loading the file as little as possible and sharing the same memory space between all VUs. This approach reduces the risk of encountering out-of-memory errors, especially in load tests involving large files.
```

Small nit here, but also out of curiosity, what does "as little as possible" mean? Does it mean k6 is able to only load parts of the file as needed, or does it mean it can load the file fewer times than the traditional open function? And if it's the latter, would it always load the file fewer times?

Author: In practice, the file will be opened twice for the whole test run under the hood. But that's an implementation detail that users should probably not be aware of, and they should treat it as if it was opened only once (O(1) space).

Thus, "as little as possible" was a way to say, "it's optimized, and the detail doesn't matter to 99.9% of users, plus they might change in the future" 😄 But I'm happy to modify it if you have a suggestion 🙇🏻

### Notes on usage

> An important consideration when using the filesystem module is its handling of external file modifications. Once a file is loaded in k6, it behaves like a "view" over the content of the file. This means that if the underlying file is modified externally during a test, those changes won't be reflected in the loaded [File]() instance.

Reviewer 2: I guess a link in `[File]()` is missing. Also, it seems like the other files have a similar issue.

> An important consideration when using the filesystem module is its handling of external file modifications. Once a file is loaded in k6, it behaves like a "view" over the content of the file. This means that if the underlying file is modified externally during a test, those changes won't be reflected in the loaded [File](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-experimental/fs/file/) instance.

Reviewer: ```suggestion
An important consideration when using the filesystem module is its handling of external file modifications. Once k6 loads a file, it behaves like a "view" over its contents. If you modify the underlying file during a test, k6 will not reflect those changes in the loaded [File](https://grafana.com/docs/k6/<K6_VERSION>/javascript-api/k6-experimental/fs/file/) instance.
```
